### PR TITLE
Indent steps for keywords like 'And'

### PR DESCRIPTION
**Pull request by rbcasperson**
The idea here is that it will help visually separate out the `Given`, `When`, and `Then` sections.

This PR isn't perfect, but I was playing around with it and got something to kind of work. Mostly at this point, I want to see how well the idea is received, though there is a certain manager type who explicitly requested it.

The current implementation works only as expected when using a different theme. When I try it with `html_theme = 'sphinx_rtd_theme'`, then for some reason the non-indented lines get weirdly highlighted. It works with `html_theme = 'classic'`.

### CODE REVIEW

**Comment by brolewis** [comment]
@rbcasperson Don't forget a version bump

**Comment by dgou** [comment]
Not yet assigned for reviewers?

**Comment by dgou** [comment]
Yes, RTD and rST have some weird interactions.
I'm curious how this would look. Lewis and I have been talking about doing some other cleanups as well, and I wonder if we need to enhance the self-tests so that they create some finalized docs to look at so a PR can post a link for review-of-how-it-looks

**Comment by dgou** [comment]
So I get the desire for visual deliniation of the G/W/T "parts" of a scenario, but the more I have been thinking about it, the less it feels like the right solution; that is, the additional steps are just as much a part of the scenario as the G/W/T tagged steps, I think the overall goal is quite valid, just not sure this is the right solution...

**Comment by dgou** [comment]
More half-baked ideas, using fixed formatting (Yay github PR review comments!)
```
* Given some initial condition
  And some other condition
* When the director yells "Places!"
  And the stagemanager yells "Lights!"
* Then the actors are all in their places
  And they are fully lit.
```

That is, using a left-margin-y way to delinate the groups without making any of the additional steps in the group indented (which to me implies they're less important. In many cases, you just have to choose an order, so that is what bothers me about the indentation.

**Comment by rbcasperson** [comment]
Thanks for the comments! Yeah I just wanted to throw this idea out there in a PR, but wasn't expected it to be merged without fair discussion. My PR assignment/documentation was lacking for sure.

Yeah I agree that I don't really like diminishing the value of the `And` steps. They are just consequentially `And`s because something else happened to come before them. I like the idea of the consistent alignment, but a marker to show the "sections" of the scenario.

**Comment by dgou** [comment]
I think it would help if we set up a doc builder for this repo (and I think we already have `.feature` files to use since @brolewis set up self tests!, so we can more easily explore aesthetic changes as well as functional ones...

**Comment by rbcasperson** [comment]
@dgou I made a new commit for a different approach that I think is closer, but I like your idea of essentially removing the bullets for the `And` steps. I just didn't know how to make that happen easily. Perhaps a bullet list isn't the right format for the steps.

Anyway, I wanted to share the next piece I tried. Here's a screenshot (not a great one):
Screen Shot 2019-07-16 at 5 02 54 PM

I don't have much bandwidth right now to try to figure out the disappearing bullet trick, but I like the idea.

**Comment by dgou** [comment]
Not able to test it out right this second, but I think if you just remove the leading dash and replace it with a space?  I hate rST except when I don't, and I'm not sure there is anything better.

**Comment by dgou** [comment]
Nah, just omitting the dash will allow Sphinx/rST to wrap the other lines.
I think using vertical bars instead would work:
```
| * Given some initial state
|   And some other initial state
| * When an action is taken
|   And another action is taken
| * Then a new state is achieved
|   And a log record is emitted
```
http://docutils.sourceforge.net/docs/user/rst/quickref.html#line-blocks

**Comment by rbcasperson** [comment]
Yes @dgou you're right on the latest commit. Pipes helped to bring us closer certainly. Here's what the latest commit makes:
Screen Shot 2019-07-17 at 8 21 42 AM

I still like the bold on the main keywords, though perhaps bold on all the keywords is a good idea. I like the dashes only on the main keywords, and as mentioned, I like the slight indent. Since `And` and `But` are shorter, the ends of the keywords just about line up.

File: sphinx_gherkindoc/writer.py [hunk 1]
```diff
@@ -193,7 +193,9 @@ def steps(steps: List[behave.model.Step]) -> None:
                 step.line,
             )
             bold_step = re.sub(r"(\\\<.*?\>)", r"**\1**", rst_escape(step.name))
-            step_text = u"- {} {}".format(_format_keyword(step.keyword), bold_step)
+            step_text = u"| - {} {}".format(_format_keyword(step.keyword), bold_step)
+            if step.keyword in ADD_ON_STEP_KEYWORDS:
+                step_text = step_text.replace("| -", "| ")
```
Comment (dgou) [inline sphinx_gherkindoc/writer.py]: If we decide to go this way, you could just make this:
```
prefix = "|  " if step.keyword in ADD... else "| -"
```
and then just `.format` it in place.
Very minor thang, just popped out at me. :-)

**Comment by dgou** [comment]
Personally I do not like bolding the keywords, it makes them stand out as more important than what the step is actually about. Maybe italics would be better?

File: sphinx_gherkindoc/writer.py (hunk 1 above)
Comment (rcasperson-jc) [reply sphinx_gherkindoc/writer.py]: Yeah definitely I'm sure there's a nicer way to write this. Once we get a visual we like, I'll clean up the code.

**Comment by rcasperson-jc** [comment]
Yeah I can understand the don't bold keywords, so I'd be fine dropping that. How about the indentation/dash removal? It's not perfectly aligned, but I do think it is a nice, fairly subtle improvement.

**Comment by dgou** [comment]
Yeah, I really really like the dash prefix, it helps, without being distracting.

**Comment by rbcasperson** [comment]
Screen Shot 2019-07-19 at 2 20 00 PM

**Comment by rbcasperson** [comment]
If it helps, I've built some docs using these changes and all works and looks good.

**Comment by dgou** [comment]
PR #10 submitted. I'd like to see that land and have this rebased on it before this one lands.

**Comment by brolewis** [comment]
So I _tried_ to be helpful and resolve the `pyproject.toml` conflict (sorry) but came across the TravisCI error. Here's the gist for resolving it (since I've already done enough damage for the day) https://gist.github.com/brolewis/5e553b49c541bc902d2ffc57f3209303

**Comment by brolewis** [comment]
image
Here's a Big Bertha scenario with `And` as well as inline-text and an inline-table.

**Comment by dgou** [comment]
Pulled this PR again (git automerged in master), when I try to build the docs, I get:
```
reading sources... [100%] tests.tags.feature-file                                                                                                                                 

Warning, treated as error:
/home/dwp/code/sphinx_gherkindoc/_docs/tests.basic.feature-file.rst:61:Block quote ends without a blank line; unexpected unindent.
Traceback (most recent call last):
  File "./bin/build_docs.py", line 107, in <module>
    main()
  File "./bin/build_docs.py", line 102, in main
    cwd=BASE_DIR,
  File "/home/linuxbrew/.linuxbrew/opt/python/lib/python3.7/subprocess.py", line 347, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['poetry', 'run', 'sphinx-build', '-c', '_docs', '-aEW', '_docs', 'docs']' returned non-zero exit status 2.
```

Also, the self-tests don't run, but there is no point in fixing those until we are agreed on the formating we want from this PR. :-)

**Comment by brolewis** [comment]
@dgou i only provided the diff needed in https://github.com/jolly-good-toolbelt/sphinx_gherkindoc/pull/3#issuecomment-513463507; it hasn't been applied to the branch yet.

**Comment by rbcasperson** [comment]
Cool the CI found a real issue with building tests! Good work Doug adding that in!

**Comment by sdutton-personal** [comment]
Piping in to say that I like the new format proposed, I haven't reviewed the code yet (not sure if its ready, or if we are still agreeing on the format).

**Comment by rbcasperson** [comment]
@sdutton-personal and all, the code is ready. There are additional test cases that demonstrate the new functionality, so take a look at those.

Based on the slack discussions about the text and tables, I decided to go the easiest route, which is to just ignore indentation when there is a table or text in _any_ step in the scenario. So this PR only changes typical scenarios that have neither. Pull down this PR and build the docs to see what it looks like.

**Pull request merged**